src/odom_node.cpp

```diff
@@ -0,0 +1,128 @@
+#include "odom_node.hpp"
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2_ros/transform_broadcaster.h>
+#include <cmath>
+#include "std_msgs/Int16.h"
+#include <geometry_msgs/PoseStamped.h>
+
+using namespace std;
+
+nav_msgs::Odometry odomNew;
+nav_msgs::Odometry odomOld;
+ 
+// Initial pose
+const double initialX = 0.0;
+const double initialY = 0.0;
+const double initialTheta = 0.00000000001;
+const double PI = 3.141592;
+ 
+// Robot physical constants
+const double TICKS_PER_REVOLUTION = 19456; // For reference purposes.
+const double wheel_distance = 0.10470; // Center of left tire to center of right tire
+const double wheel_conversion_left = 0.13194 / 19456.0; // perimeter over pulses per rev
+const double wheel_conversion_right = 0.13194 / 19456.0;
+
+// Encoder value
+int pos_left = 0;
+int pos_right = 0;
+int pos_left_prev = 0;
+int pos_right_prev = 0;
+
+
+ 
+// Publish a nav_msgs::Odometry message in quaternion format
+void publish_quat(ros::Publisher &odom_data_pub_quat) {
+ 
+  tf2::Quaternion q;
+         
+  q.setRPY(0, 0, odomNew.pose.pose.orientation.z);
+ 
+  nav_msgs::Odometry quatOdom;
+  quatOdom.header.stamp = odomNew.header.stamp;
+  quatOdom.header.frame_id = "odom";
+  quatOdom.child_frame_id = "base_link";
+  quatOdom.pose.pose.position.x = odomNew.pose.pose.position.x;
+  quatOdom.pose.pose.position.y = odomNew.pose.pose.position.y;
+  quatOdom.pose.pose.position.z = odomNew.pose.pose.position.z;
+  quatOdom.pose.pose.orientation.x = q.x();
+  quatOdom.pose.pose.orientation.y = q.y();
+  quatOdom.pose.pose.orientation.z = q.z();
+  quatOdom.pose.pose.orientation.w = q.w();
+  quatOdom.twist.twist.linear.x = odomNew.twist.twist.linear.x;
+  quatOdom.twist.twist.linear.y = odomNew.twist.twist.linear.y;
+  quatOdom.twist.twist.linear.z = odomNew.twist.twist.linear.z;
+  quatOdom.twist.twist.angular.x = odomNew.twist.twist.angular.x;
+  quatOdom.twist.twist.angular.y = odomNew.twist.twist.angular.y;
+  quatOdom.twist.twist.angular.z = odomNew.twist.twist.angular.z;
+ 
+  for(int i = 0; i<36; i++) {
```

Comment:
Comment to explain this?

Reply:
Covariance matrix is 6x6, loop through to set diagonal values